test/demo.jl

```diff
@@ -42,15 +48,31 @@ end
 @testset "Demo ComponentsArrays" begin
 
     # target distribution parametrized by ComponentsArray
-    p1 = ComponentVector(μ = 2.0, σ=1)
-    ℓπ(p::ComponentArray) = -(1 .- p.μ)^2/(p.σ)^2
+    p1 = ComponentVector(μ=2.0, σ=1)
+    struct DemoProblemComponentArrays end
+    function LogDensityProblems.logdensity(::DemoProblemComponentArrays, p::ComponentArray)
+        return -((1 - p.μ) / p.σ)^2
+    end
+    LogDensityProblems.dimension(::DemoProblemComponentArrays) = 2
+    ℓπ = DemoProblemComponentArrays()
 
     # Define a Hamiltonian system
     D = length(p1)          # number of parameters
     metric = DiagEuclideanMetric(D)
 
     # choose AD framework or provide a function manually
-    hamiltonian = Hamiltonian(metric, ℓπ, ForwardDiff)
+    ℓπ_with_grad = LogDensityProblemsAD.ADgradient(
+        Val(:ForwardDiff),
+        ℓπ;
+        # NOTE: Have to manually construct the `gradientconfig` to ensure we're not using the default
+        # buffer in the config (which won't be of type `ComponentArray`).
+        gradientconfig = ForwardDiff.GradientConfig(
+            Base.Fix1(LogDensityProblems.logdensity, ℓπ),
+            similar(p1),
+            LogDensityProblemsAD._default_chunk(ℓπ),
+        )
+    )
+    hamiltonian = Hamiltonian(metric, AbstractMCMC.LogDensityModel(ℓπ_with_grad))
```

Comment:
Requires LogDensityProblemsAD 1.1.0 but only in the tests it seems:
```suggestion
    hamiltonian = Hamiltonian(metric, ℓπ, Val(:ForwardDiff); x = p1)
```